File: box/src/main/java/de/qabel/box/storage/AbstractMetadata.java
package de.qabel.box.storage;

import de.qabel.box.storage.exceptions.QblStorageCorruptMetadata;
import de.qabel.box.storage.exceptions.QblStorageException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public abstract class AbstractMetadata {
    public static final int TYPE_NONE = -1;
    protected static final Logger logger = LoggerFactory.getLogger(DirectoryMetadata.class);
    protected static final String JDBC_PREFIX = "jdbc:sqlite:";
    protected final Connection connection;
    protected File tempDir;
    File path;

    public AbstractMetadata(Connection connection, File path) {
        this.connection = connection;
        this.path = path;
    }

    /**
     * Path of the metadata file on the local filesystem
     */
    public File getPath() {
        return path;
    }

    /**
     * Writable temporary directory which is used for encryption and decryption
     */
    public File getTempDir() {
        return tempDir;
    }

    Integer getSpecVersion() throws QblStorageException {
        try (Statement statement = connection.createStatement()) {
            try (ResultSet rs = statement.executeQuery(
                "SELECT version FROM spec_version")) {
                if (rs.next()) {
                    return rs.getInt(1);
                } else {
                    throw new QblStorageCorruptMetadata("No version found!");
                }
            }
        } catch (SQLException e) {
            throw new QblStorageException(e);
        }
    }

    protected void initDatabase() throws SQLException, QblStorageException {
        for (String q : getInitSql()) {
            try (Statement statement = connection.createStatement()) {
                statement.executeUpdate(q);
            }
        }
    }

    protected abstract String[] getInitSql();
}

File: box/src/main/java/de/qabel/box/storage/AbstractMetadata.kt
package de.qabel.box.storage

import de.qabel.box.storage.exceptions.QblStorageCorruptMetadata
import de.qabel.box.storage.exceptions.QblStorageException
import de.qabel.core.repository.sqlite.ClientDatabase
import de.qabel.core.repository.sqlite.tryWith
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.sql.PreparedStatement
import java.sql.SQLException

abstract class AbstractMetadata(val connection: ClientDatabase, path: File) {
    val logger: Logger by lazy { LoggerFactory.getLogger(AbstractMetadata::class.java) }

    /**
     * Path of the metadata file on the local filesystem
     */
    var path: File
        internal set

    init {
        this.path = path
    }

    @JvmName("getSpecVersion")
    fun findSpecVersion(): Int? =
        try {
            tryWith(connection.prepare("SELECT version FROM spec_version")) {
                tryWith(executeQuery()) {
                    if (next()) {
                        return getInt(1)
                    } else {
                        throw QblStorageCorruptMetadata("No version found!")
                    }
                }
            }
        } catch (e: SQLException) {
            throw QblStorageException(e)
        }

    @Throws(QblStorageException::class)
    protected fun executeStatement(statementCallable: () -> PreparedStatement) {
        try {
            tryWith(statementCallable()) {
                val changedLines = executeUpdate()
                if (changedLines != 1) {
                    throw QblStorageException("Failed to execute statement, changedLines $changedLines != 1")
                }
            }
        } catch (e: Exception) {
            throw QblStorageException(e)
        }
    }

    companion object {
        val TYPE_NONE = -1
        @JvmField
        val DEFAULT_JDBC_PREFIX: String = "jdbc:sqlite:"
    }
}

